### PR TITLE
fix(19577): fix regression with fully inferred types and non-null assertions

### LINKED ISSUE #19577: Regression: Not-null assertion causes implicit any
**TypeScript Version:**  2.7.0-dev.20171029

**Code**

```ts
function test(numbers: number[]) {
    let last;

    for (const n of numbers) {
        if (n % 2) {
            return n;
        }

        last = n;
    }

    return last!; // without the bang, last is inferred as `number | null`
                  // adding the bang causes implicit any
}
```

**Workaround:**
Initialise the "tracking" variable with undefined.

```ts
function test(numbers: number[]) {
    let last = undefined;

    for (const n of numbers) {
        if (n % 2) {
            return n;
        }

        last = n;
    }

    return last!; // OK
}
```

**Expected behavior:**
Should compile successfully with and without a bang.

### PR DESCRIPTION
**Pull request by nicolas377**
Fixes #19577

### CODE REVIEW

File: tests/cases/compiler/nonNullFullInference.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+        if (n % 2) {
+            return n;
+        }
+
+        last = n;
+    }
+
+    last; // number
+    last!; // number
+}
```
Comment (nicolas377) [inline tests/cases/compiler/nonNullFullInference.ts:17]: I feel like this isn't enough tests. Does anyone have suggestions for tests to add?

File: tests/baselines/reference/nonNullFullInference.types [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+        }
+
+        last = n;
+>last = n : number
+>last : any
+>n : number
+    }
+
+    last; // number
+>last : number
```
Comment (Andarist) [inline tests/baselines/reference/nonNullFullInference.types:31]: Is this correct though? `numbers` could always be an empty array so `last` could have never been assigned to. The current inferred type for this is `number | undefined` and it reflects the runtime behavior better.

File: tests/cases/compiler/nonNullFullInference.ts (hunk 1 above)
Comment (Andarist) [reply tests/cases/compiler/nonNullFullInference.ts:17]: I'm not sure if this is any different from the current test case - there is a good chance that it isn't - but I've seen that your changes are also related to `autoArrayType` so I've tried to use it in a test case. 

https://www.typescriptlang.org/play?ts=4.7.4#code/GYVwdgxgLglg9mABFApgZygOQZkAbPASTGBQCcVIUAKMEAWwCNy0AuROp8gbQF0BKRAG8AUInGI8KKIgpp8UMRIgIMiAIZkyiALyI+IpeOBxt1FWDVI4wDg2Zk0g0RNeIYt2ogCkiAEzORm6uFFAgZEjcYLwA3EGuAL6GweKaZAB0AA4gaAAWtPzxEnIKuhpaQUlBJXhQca4A9A2IAHoA-NXoCgCEQU2tHZ1hEbJdtSIJQA

File: tests/baselines/reference/nonNullFullInference.types (hunk 2 above)
Comment (nicolas377) [reply tests/baselines/reference/nonNullFullInference.types]: It seems either the baseline generation omits `undefined` in the types, or this existed prior to the change being made.

**Comment by weswigham** [comment]
@typescript-bot pack this
@typescript-bot user test this inline
@typescript-bot test this
@typescript-bot run dt
@typescript-bot perf test this

**Review by weswigham — APPROVED** [review]
Honestly, the tests came back clean and the change seems somewhat reasonable - prioritizing the `undefined` starting type for a non-null-asserted auto type seems fair, as does getting the nonnullable type of the flow type (though I feel like the containing non-null assertion expression should maybe be doing that, but I guess you have to in order to avoid a spurious use before def error for auto-typed things).

This just needs to be updated for current `main` and then DT run again as a quick double check that it's still OK.

**Comment by gabritto** [comment]
@typescript-bot pack this

**Pull request merged**